blog/2024-08-21-amqp-benchmarks/index.md

```diff
@@ -0,0 +1,677 @@
+
+RabbitMQ flow control settings present a trade-off:
+* Low values ensure stability in production.
+* High values can result in higher performance for individual connections but may lead to higher memory spikes when many connections publish large messages concurrently.
+
+RabbitMQ uses conservative flow control default settings to favour stability in production over winning performance benchmarks.
+
+### AMQP 1.0 in 4.0
+
+```
+# quiver //host.docker.internal//queues/my-quorum-queue \
+    --durable --count 1m --duration 10m --body-size 12 --credit 5000
+
+RESULTS
+
+Count ............................................. 1,000,000 messages
+Duration ............................................... 12.0 seconds
+Sender rate .......................................... 83,459 messages/s
+Receiver rate ........................................ 83,396 messages/s
+End-to-end rate ...................................... 83,181 messages/s
+
+Latencies by percentile:
+
+          0% ........ 9 ms       90.00% ....... 47 ms
+         25% ....... 27 ms       99.00% ....... 61 ms
+         50% ....... 35 ms       99.90% ....... 76 ms
+        100% ....... 81 ms       99.99% ....... 81 ms
+```
+
+<details>
+<summary>Default Flow Control Settings</summary>
+
+The previous benchmark calls fsync 1,244 times in module [ra_log_wal](https://github.com/rabbitmq/ra/blob/e95ab7b9df1f8f4ffec8535d60185b3bc33a09bc/src/ra_log_wal.erl#L770) (Raft write ahead log).
+
+The same benchmark with default flow control settings calls fsync 15,493 times resulting in significantly lower throughput:
+```
+# quiver //host.docker.internal//queues/my-quorum-queue \
+    --durable --count 1m --duration 10m --body-size 12 --credit 5000
+
+RESULTS
+
+Count ............................................. 1,000,000 messages
+Duration .............................................. 100.2 seconds
+Sender rate ........................................... 9,986 messages/s
+Receiver rate ......................................... 9,987 messages/s
+End-to-end rate ....................................... 9,983 messages/s
+
+Latencies by percentile:
+
+          0% ....... 10 ms       90.00% ....... 24 ms
+         25% ....... 14 ms       99.00% ....... 30 ms
+         50% ....... 18 ms       99.90% ....... 38 ms
+        100% ....... 55 ms       99.99% ....... 47 ms
+```
+
+Each fsync took 5.9 ms on average.
+
+```
+(15,493 - 1,244) * 5.9 ms = 84 seconds
+```
+
+Therefore, this benchmark with default flow control settings is blocked for 84 seconds longer executing fsync than the previous benchmark with increased flow control settings.
+This shows how critical enterprise-grade high performance disks are to get the best results out of quorum queues.
+For your production workloads, we recommend using disks that fsync faster rather than increasing RabbitMQ flow control settings.
+
+It's worth noting that the `ra_log_wal` is shared across all quorum queues on a given RabbitMQ node.
```

Comment:
Should this be "Raft log WAL"? I don't think most readers would guess that this is a module name.